internal/core/src/mmap/ChunkData.h

```diff
@@ -0,0 +1,200 @@
+        return data_[i];
+    }
+    const ChunkViewType<Type>&
+    operator[](const int i) const {
+        return view(i);
+    }
+    void*
+    data() {
+        return data_.data();
+    };
+    size_t
+    size() {
+        return size_;
+    };
+
+ private:
+    int64_t size_ = 0;
+    FixedVector<ChunkViewType<Type>> data_;
+    storage::MmapChunkDescriptor mmap_descriptor_ = std::nullopt;
+};
+template <>
+inline void
+VariableLengthChunk<std::string>::set(const std::string* src,
+                                      uint32_t begin,
+                                      uint32_t length) {
+    auto mcm = storage::MmapManager::GetInstance().GetMmapChunkManager();
+    milvus::ErrorCode err_code;
+    AssertInfo(
+        begin + length <= size_,
+        "failed to set a chunk with length: {} from beign {}, map_size={}",
+        length,
+        begin,
+        size_);
+    for (auto i = 0; i < length; i++) {
+        auto buf_size = src[i].size() + 1;
+        auto buf = (char*)mcm->Allocate(mmap_descriptor_, buf_size);
+        AssertInfo(buf != nullptr,
+                   "failed to allocate memory from mmap_manager, error_code");
+        std::strcpy(buf, src[i].c_str());
+        data_[i + begin] = std::string_view(buf, src[i].size());
+    }
+}
+template <>
+inline std::string
+VariableLengthChunk<std::string>::get(const int i) const {
+    // copy to a string
+    return std::string(data_[i]);
+}
+template <>
+inline void
+VariableLengthChunk<Json>::set(const Json* src,
+                               uint32_t begin,
+                               uint32_t length) {
+    auto mcm = storage::MmapManager::GetInstance().GetMmapChunkManager();
+    milvus::ErrorCode err_code;
+    AssertInfo(
+        begin + length <= size_,
+        "failed to set a chunk with length: {} from beign {}, map_size={}",
+        length,
+        begin,
+        size_);
+    for (auto i = 0; i < length; i++) {
+        auto buf_size = src[i].size() + simdjson::SIMDJSON_PADDING + 1;
+        auto buf = (char*)mcm->Allocate(mmap_descriptor_, buf_size);
+        AssertInfo(
+            buf != nullptr,
+            "failed to allocate memory from mmap_manager, error_code:{}");
+        std::strcpy(buf, src[i].c_str());
+        data_[i + begin] = Json(buf, src[i].size());
+    }
+}
+template <>
+inline Json
+VariableLengthChunk<Json>::get(const int i) const {
+    return std::move(Json(simdjson::padded_string(data_[i].data())));
+}
+template <>
+inline void
+VariableLengthChunk<Array>::set(const Array* src,
+                                uint32_t begin,
+                                uint32_t length) {
+    auto mcm = storage::MmapManager::GetInstance().GetMmapChunkManager();
+    milvus::ErrorCode err_code;
+    AssertInfo(
+        begin + length <= size_,
+        "failed to set a chunk with length: {} from beign {}, map_size={}",
+        length,
+        begin,
+        size_);
+    for (auto i = 0; i < length; i++) {
```

Comment:
allocate once to make sure all data is in order and no small chunks

Reply:
do this job in next pr